add generation time metrics (#613)
openvinotoolkit/openvino.genai: src/README.md

}
```

> ### Performance Metrics

Reviewer: When it gets merged, please, open another PR adding it to C++ and Python docstrings.

Author: Done https://github.com/openvinotoolkit/openvino.genai/pull/713

> `ov.genai.PerfMetrics` (referred to as `PerfMetrics` for simplicity) is a structure that holds performance metrics for each generate call. `PerfMetrics` hold fields with mean and standard deviations for the following metrics:

Reviewer: There's
> `PerfMetrics` has

below. Like `PerfMetrics` is singular. But here
> `PerfMetrics` hold

`hold` doesn't have `s`, like if `PerfMetrics` were plural. Bot of the approaches seem to be valid, but you need to stick to one of them.

Author: agreed. Sticked to the plural

> `ov.genai.PerfMetrics` (referred to as `PerfMetrics` for simplicity) is a structure that holds performance metrics for each generate call. `PerfMetrics` hold fields with mean and standard deviations for the following metrics:

Author: agreed

- `ttft`
- `tpot`

> - `load_time`

Author: agreed, moved this metric below because it does not have mean/std. Added dimentions as welll

- `generate_duration`
- `tokenization_duration`
- `detokenization_duration`
- `throughput`

and:
- `num_generated_tokens`

> - `num_input_tokens`

Reviewer: Please define each of the metrics. For example, when I first saw `tpot`, I though it was misspelled tput standing for throughput.

Author: defined metrics/dimensions

> - `num_input_tokens`

Author: They are left because one corresponds to the number of tokens in the initial input promt and tokenization time, other to the number of generated tokens and detokenization time.

If we need them both we always can sum them. But if is do the sum on our side if someone want to separate them they will have to do extra efforts, e.g. trying to tokenize input prompt one again. I think it's better to leave separate

Performance metrics are stored either in the `DecodedResults` or `EncodedResults` `perf_metric` field. Additionally to the fields mentioned above, `PerfMetrics` has a member `raw_metrics` of type `ov.genai.RawPerfMetrics` (referred to as `RawPerfMetrics` for simplicity) that contains raw values for the durations of each batch of new token generation, tokenization durations, detokenization durations, and more. These raw metrics are accessible if you wish to calculate your own statistical values such as median or percentiles. However, since mean and standard deviation values are usually sufficient, we will focus on `PerfMetrics`.

```python
import openvino_genai as ov_genai
pipe = ov_genai.LLMPipeline(model_path, "CPU")
res = pipe.generate(["The Sun is yellow because"], max_new_tokens=20)
perf_metrics = res.perf_metrics
print(f'generate_duration: {perf_metrics.mean_generate_duration:.2f}')
print(f'ttft: {perf_metrics.mean_ttft:.2f}')
print(f'tpot: {perf_metrics.mean_tpot:.2f}')
```
output:
```sh
mean_generate_duration: 76.28
mean_ttft: 42.58
mean_tpot 3.80
```

> >**Note**: If the input prompt is just a string, the generate function will return only a string without perf_metrics. To obtain perf_metrics, provide the prompt as a list with at least one element or call generate with encoded inputs.

Reviewer: ```suggestion
>**Note**: If the input prompt is just a string, the generate function returns only a string without perf_metrics. To obtain perf_metrics, provide the prompt as a list with at least one element or call generate with encoded inputs.
```

Author: agreed

> Several `perf_metrics` can be added with each other. In that case `raw_metrics` will be concatenated and mean/std values will be recalculated. This enhances benchmarking and accumulating statistics from several calls.

Reviewer: ```suggestion
Several `perf_metrics` can be added to each other. In that case `raw_metrics` are concatenated and mean/std values are recalculated. This accumulates statistics from several `generate()` calls.
```

Author: agreed

```python
import openvino_genai as ov_genai
pipe = ov_genai.LLMPipeline(model_path, "CPU")
res_1 = pipe.generate(["The Sun is yellow because"], max_new_tokens=20)
res_2 = pipe.generate(["Why Sky is blue because"], max_new_tokens=20)
perf_metrics = res_1.perf_metrics + res_2.perf_metrics

print(f'generate_duration: {perf_metrics.mean_generate_duration:.2f}')
print(f'ttft: {perf_metrics.mean_ttft:.2f}')

> print(f'tpot: {perf_metrics.mean_tpot:.2f}')

Reviewer: If you are going to provide snippets only for one language, it should be C++ because it's mor verbose and one of the GenAI features is the exitance of C++ interface.
The better option would be to cover both languages.
An even better option would be not to have them at all because there's no easy automatic way to validate the snippets. Refer to samples.

Author: Added to c++ as well.

```
An even better option would be not to have them at all because there's no easy automatic way to validate the snippets. Refer to samples.
```
probably it's possible to parse md extract snippets and compile them, but id doesn't seem trivial.

I referred to samples for more detailed usage.
